### PR TITLE
Task2:Date Server and Client

### PR DESCRIPTION
**Pull request by kristiyanpetkov**

### CODE REVIEW

File: src/com/clouway/dateserverandclient/Client.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    private final String serverIP;
+    private final int serverPort;
+
+    public Client(String serverIP, int serverPort, Display display) {
+        this.serverIP = serverIP;
+        this.serverPort = serverPort;
+        this.display=display;
+    }
+
+    public String connectClient() {
```
Comment (mgenov) [inline src/com/clouway/dateserverandclient/Client.java]: I don't get why you are having too much class level references which are pointing to method level instances which are going to be garbage collected after method connectClient is invoked. I would like these design decisions to be considered at writing of the code.   

Also there is a smell in the design of this method. If you are showing message to the display, why you are returning the same message?

File: src/com/clouway/dateserverandclient/DateServer.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    private ServerSocket serverSocket;
+    private Socket socket;
+
+    public void startServer() {
+        (thread=new Thread() {
+            @Override
+            public void run() {
+                try {
```
Comment (mgenov) [inline src/com/clouway/dateserverandclient/DateServer.java]: Please use smaller scope for the try/catch block.

File: test/com/clouway/dateserverandclienttest/DateServerTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery() {{
+        setThreadingPolicy(new Synchroniser());
+    }};
+
+    @Mock
+    Clock clock;
+
+    private void pretendThatCurrentDateIs(final Date date) {
```
Comment (mgenov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Private method should be defined after public methods. 

https://github.com/clouway/teamspace/wiki/Java%20Code%20Convention

File: src/com/clouway/dateserverandclient/Client.java (hunk 1 above)
Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/Client.java]: I was making some experiments and i forgot to remove the unnecessary class level references. I already fixed that. Code smell also removed. Now i only show message to the display and assert this message in the Test if it is the same with the current date.

Comment (mgenov) [inline src/com/clouway/dateserverandclient/Client.java]: One suggestions: please check your code in PR before you assign me as reviewer.

File: src/com/clouway/dateserverandclient/DateServer.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+/**
+ * Created by clouway on 15-12-21.
+ */
+public class DateServer {
+    private final int port;
+    private final Clock clock;
+    private Thread thread;
```
Comment (mgenov) [inline src/com/clouway/dateserverandclient/DateServer.java]: Why you need this thread reference when you are not using it ?

File: src/com/clouway/dateserverandclient/DateServer.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    public DateServer(int port, Clock clock) {
+        this.port = port;
+        this.clock = clock;
+    }
+
+    private ServerSocket serverSocket;
+    private Socket socket;
```
Comment (mgenov) [inline src/com/clouway/dateserverandclient/DateServer.java]: Why you are holding socket reference to the last connected client ? What happens when client did not close the socket and waits the server to make it ?

File: test/com/clouway/dateserverandclienttest/DateServerTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.Date;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by clouway on 15-12-21.
+ */
+public class DateServerTest {
+    public class DisplayDate implements Display {
```
Comment (mgenov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Why DisplayDate ?

File: test/com/clouway/dateserverandclienttest/DateServerTest.java (hunk 3 above)
Comment (kristiyanpetkov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: By now i didn't know about the proper place of the methods. I find the link about Code Convention very useful.

File: src/com/clouway/dateserverandclient/DateServer.java (hunk 4 above)
Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/DateServer.java]: i used it before in the method stopServer() and i needed thread reference. I forgot to fix it. Already did.

File: test/com/clouway/dateserverandclienttest/DateServerTest.java (hunk 6 above)
Comment (kristiyanpetkov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: i think to rename it to DisplayMessage. Is it better name?

Comment (mgenov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Why DisplayMessage ?

File: src/com/clouway/dateserverandclient/DateServer.java (hunk 2 above)
Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/DateServer.java]: I split it in two try/catch blocks.

File: test/com/clouway/dateserverandclienttest/DateServerTest.java (hunk 6 above)
Comment (kristiyanpetkov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: It does not have to be common. It should describe what happens inside this class. Maybe DisplayCurrentDateAndTime will be better.

Comment (mgenov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Or just simple FakeDisplay would be simpler and will say that this is used for testing purposes.

File: src/com/clouway/dateserverandclient/Client.java (hunk 1 above)
Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/Client.java]: Okay. Hopefully this was the last time.

File: test/com/clouway/dateserverandclienttest/DateServerTest.java (hunk 6 above)
Comment (mgenov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Name of objects should be noun, not verb. 

Ruby developers has style like: DisplaysMessages which could be used too, but it's different from suggested at the moment.

Comment (kristiyanpetkov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Yes, It is actually the fake implementation in my test class, so i will rename it to FakeDisplay.

File: src/com/clouway/dateserverandclient/DateServer.java (hunk 5 above)
Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/DateServer.java]: what i read and understood is that when the server is closed by invoking .close() method , the client recieve FIN flag which tells that the server closed the connection and the operation system will close all sockets connected with that process.

Comment (mgenov) [inline src/com/clouway/dateserverandclient/DateServer.java]: This doesn't answer the question why you keep reference to the last connected client.

Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/DateServer.java]: I used that reference in the close method. After i read some more documentation i know it is not correct to keep that reference and to use it in the close method.

File: src/com/clouway/dateserverandclient/DateServer.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                        e.printStackTrace();
+                    }
+                }
+            }
+        }).start();
+    }
+
+    public void stopServer() {
+        try {
+            serverSocket.close();
```
Comment (mgenov) [inline src/com/clouway/dateserverandclient/DateServer.java]: How thread will exit when you stop the server socket ?

Comment (kristiyanpetkov) [inline src/com/clouway/dateserverandclient/DateServer.java]: it doesn't stop and remain as a process in the memory. i fixed it.

Comment (mgenov) [inline src/com/clouway/dateserverandclient/DateServer.java]: It's thread, it's not a process. 

Actually you can use the exception to do that: 
```java
loop() {
  try {
  
  }catch(IOException e) {
     break;
  }
}

```
which means that run method will exit and thread will stop.

File: test/com/clouway/dateserverandclienttest/DateServerTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        @Override
+        public void setMessage(String message) {
+            this.message = message;
+            System.out.println(message);
+        }
+    }
+
+
+        @Rule
```
Comment (mgenov) [inline test/com/clouway/dateserverandclienttest/DateServerTest.java]: Code is not formatted.

**Comment by mgenov** [comment]
LGTM !

**Pull request merged**